### PR TITLE
Workaround to fix eventloop.shutdownGracefully hanging issue

### PR DESCRIPTION
**Pull request by zoewangg**
## Description
Update `NettyNioAsyncHttpClient#close` to wait for `pools#close` before shutting down eventloop

The root cause is a race condition in Netty where the event loop gets terminated before channel is closed so `globalEventExecutor` will be stuck forever waiting for eventloop to notify the `channel.close` future.

```
"globalEventExecutor-1-1" #62 prio=5 os_prio=31 tid=0x00007ff34abb8800 nid=0x8c0b in Object.wait() [0x0000700003023000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x000000076b1b6a30> (a io.netty.channel.DefaultChannelPromise)
	at java.lang.Object.wait(Object.java:502)
	at io.netty.util.concurrent.DefaultPromise.awaitUninterruptibly(DefaultPromise.java:253)
	- locked <0x000000076b1b6a30> (a io.netty.channel.DefaultChannelPromise)
	at io.netty.channel.DefaultChannelPromise.awaitUninterruptibly(DefaultChannelPromise.java:137)
	at io.netty.channel.DefaultChannelPromise.awaitUninterruptibly(DefaultChannelPromise.java:30)
	at io.netty.channel.pool.SimpleChannelPool.close(SimpleChannelPool.java:398)
	at software.amazon.awssdk.http.nio.netty.internal.HonorCloseOnReleaseChannelPool.close(HonorCloseOnReleaseChannelPool.java:75)
	at software.amazon.awssdk.http.nio.netty.internal.utils.BetterFixedChannelPool$6.run(BetterFixedChannelPool.java:382)
	at io.netty.util.concurrent.GlobalEventExecutor$TaskRunner.run(GlobalEventExecutor.java:248)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
```

See https://github.com/aws/aws-sdk-java-v2/issues/1109

~~This is a workaround to wait for 2 seconds before shutting down eventloop to avoid the race condition~~

The new better workaround is creating a `BetterSimpleChannelPool` to extend `SimpleChannelPool` and provide an asynchronous close method and then make `AwaitCloseChannelPoolMap` to wait to make sure all channels are being closed before we invoke `EventLoopGroup.shutdownGracefully`

~~**NOTE**
The unit tests are still pending because I wanted to make sure this approach is okay before I spend time adding the tests.~~

## Testing
Tested with the spring-boot app I wrote to replicate the issue and verified the issue was fixed with this change.

All integration tests passed

~~TODO: add more unit tests.~~(done)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/BetterSimpleChannelPool.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import io.netty.channel.pool.SimpleChannelPool;
+import java.util.concurrent.CompletableFuture;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+
+/**
+ * Extension of {@link SimpleChannelPool} to add an asynchronous close method
+ */
+@SdkInternalApi
+public final class BetterSimpleChannelPool extends SimpleChannelPool {
+    private volatile CompletableFuture<Boolean> isClosed;
```
Comment (millems) [inline http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/BetterSimpleChannelPool.java]: Can this be final, instead of volatile?

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/AwaitCloseChannelPoolMap.java [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+        channelPoolRef.set(wrapSimpleChannelPool(bootstrap, simpleChannelPool));
+        return new SimpleChannelPoolAwareChannelPool(simpleChannelPool, channelPoolRef.get());
+    }
+
+    @Override
+    public void close() {
+        log.trace(() -> "Closing channel pools");
+        Collection<SimpleChannelPoolAwareChannelPool> channelPools = pools().values();
+        super.close();
```
Comment (millems) [inline http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/AwaitCloseChannelPoolMap.java]: Could something get added between snapshotting the channels in the pools and cleaning up?

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/NettyConfiguration.java [hunk 3]
```diff
@@ -32,6 +32,7 @@
 @SdkInternalApi
 public final class NettyConfiguration {
 
+    public static final int CHANNEL_POOL_CLOSE_TIMEOUT_SECONDS = 1;
```
Comment (millems) [inline http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/NettyConfiguration.java]: Is this long enough for really big channel pools?

File: http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClientWireMockTest.java [hunk 4]
```diff
@@ -229,19 +228,18 @@ public void customChannelFactoryIsUsed() throws Exception {
     public void closeClient_shouldCloseUnderlyingResources() {
         SdkEventLoopGroup eventLoopGroup = SdkEventLoopGroup.builder().build();
         ChannelPool channelPool = mock(ChannelPool.class);
-        SdkChannelPoolMap<URI, ChannelPool> sdkChannelPoolMap = new SdkChannelPoolMap<URI, ChannelPool>() {
+        SdkChannelPoolMap<URI, ? extends ChannelPool> sdkChannelPoolMap = new SdkChannelPoolMap<URI, ChannelPool>() {
```
Comment (millems) [inline http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClientWireMockTest.java]: Is this change needed?

File: http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/AwaitCloseChannelPoolMapTest.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.nio.netty.internal;
+
+/**
+ * Class javadoc
+ */
+public class AwaitCloseChannelPoolMapTest {
+}
```
Comment (millems) [inline http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/AwaitCloseChannelPoolMapTest.java]: Nice

File: http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClientWireMockTest.java (hunk 4 above)
Comment (zoewangg) [reply http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClientWireMockTest.java]: Nope, will remove

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/BetterSimpleChannelPool.java (hunk 1 above)
Comment (zoewangg) [reply http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/BetterSimpleChannelPool.java]: Yeah, will update

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/AwaitCloseChannelPoolMap.java (hunk 2 above)
Comment (zoewangg) [reply http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/AwaitCloseChannelPoolMap.java]: It's definitely possible but should be very unlikely to happen in my mind. I'm not very concerned about this edge case.

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/NettyConfiguration.java (hunk 3 above)
Comment (zoewangg) [reply http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/NettyConfiguration.java]: I suppose it should be really quick to close all channel pools, but I can increase it to 5 seconds.

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/AwaitCloseChannelPoolMap.java (hunk 2 above)
Comment (millems) [reply http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/AwaitCloseChannelPoolMap.java]: What's the impact in the rare case that it does?

Comment (zoewangg) [reply http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/AwaitCloseChannelPoolMap.java]: The impact is the same race condition where the event loop threads get terminated before the newly added channels being closed and it could cause the eventLoop.shutdown to hang.

Comment (millems) [reply http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/AwaitCloseChannelPoolMap.java]: Hmm. Is there any way to avoid that, or no?

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/BetterSimpleChannelPool.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        isClosed = new CompletableFuture<>();
+    }
+
+    @Override
+    public void close() {
+        super.close();
+        isClosed.complete(true);
+    }
+
+    CompletableFuture<Boolean> isClosed() {
```
[the remark this answers is not in the document]
Comment (zoewangg) [reply http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/BetterSimpleChannelPool.java]: Yeah, sounds good

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/AwaitCloseChannelPoolMap.java [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+        BetterSimpleChannelPool simpleChannelPool = new BetterSimpleChannelPool(bootstrap, handler);
+
+        channelPoolRef.set(wrapSimpleChannelPool(bootstrap, simpleChannelPool));
+        return new SimpleChannelPoolAwareChannelPool(simpleChannelPool, channelPoolRef.get());
+    }
+
+    @Override
+    public void close() {
+        log.trace(() -> "Closing channel pools");
+        // If there is a new pool being added while we are iterating the pools, exception could be thrown.
```
Comment (dagnir) [inline http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/AwaitCloseChannelPoolMap.java]: If a new pool is acquired, is there a possibility of it creating a channel that won't get closed?

Comment (zoewangg) [reply http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/AwaitCloseChannelPoolMap.java]: The new pool would be closed since we are calling `super.close()`, but there is a possibility that the same race condition could occur between the newlyAcquiredChannelPool.close and eventloop.shutdown

Comment (dagnir) [reply http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/AwaitCloseChannelPoolMap.java]: Is it possible a new pool is created while `super.close()` is running though? It doesn't look like access to the underlying map is synchronized.

Comment (zoewangg) [reply http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/AwaitCloseChannelPoolMap.java]: Yeah, it's possible. There's no way to synchronize retrieval operations with `ConcurrentHashMap` and there's going to be performance degradation if we switch to synchronized map. I'm not too concerned about this rare edge case.

Comment (dagnir) [reply http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/AwaitCloseChannelPoolMap.java]: Okay. Can we at least note it? Maybe we can find a way to fill in this gap down the line that won't affect performance.

Comment (zoewangg) [reply http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/AwaitCloseChannelPoolMap.java]: Yeah, added notes.
